ipn/ipnlocal/web_client.go

```diff
@@ -40,51 +43,54 @@ type webClient struct {
 // Specifially, it sets b.web.lc to the provided LocalClient.
 // If provided as nil, b.web.lc is cleared out.
 func (b *LocalBackend) ConfigureWebClient(lc *tailscale.LocalClient) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	b.webClient.mu.Lock()
+	defer b.webClient.mu.Unlock()
 	b.webClient.lc = lc
 }
 
-// WebClientInit initializes the web interface for managing this
-// tailscaled instance.
-// If the web interface is already running, WebClientInit is a no-op.
-func (b *LocalBackend) WebClientInit() (err error) {
+// webClientGetOrInit gets or initializes the web server for managing
+// this tailscaled instance.
+// s is always non-nil if err is empty.
+func (b *LocalBackend) webClientGetOrInit() (s *web.Server, err error) {
 	if !b.ShouldRunWebClient() {
-		return errors.New("web client not enabled for this device")
+		return nil, errors.New("web client not enabled for this device")
 	}
 
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	b.webClient.mu.Lock()
+	defer b.webClient.mu.Unlock()
 	if b.webClient.server != nil {
-		return nil
+		return b.webClient.server, nil
 	}
 
-	b.logf("WebClientInit: initializing web ui")
+	b.logf("webClientGetOrInit: initializing web ui")
 	if b.webClient.server, err = web.NewServer(web.ServerOpts{
 		Mode:        web.ManageServerMode,
 		LocalClient: b.webClient.lc,
 		Logf:        b.logf,
 	}); err != nil {
-		return fmt.Errorf("web.NewServer: %w", err)
+		return nil, fmt.Errorf("web.NewServer: %w", err)
 	}
 
-	b.logf("WebClientInit: started web ui")
-	return nil
+	b.logf("webClientGetOrInit: started web ui")
+	return b.webClient.server, nil
 }
 
 // WebClientShutdown shuts down any running b.webClient servers and
 // clears out b.webClient state (besides the b.webClient.lc field,
 // which is left untouched because required for future web startups).
 // WebClientShutdown obtains the b.mu lock.
-func (b *LocalBackend) WebClientShutdown() {
+func (b *LocalBackend) webClientShutdown() {
 	b.mu.Lock()
-	server := b.webClient.server
-	b.webClient.server = nil
 	for ap, ln := range b.webClientListeners {
 		ln.Close()
 		delete(b.webClientListeners, ap)
 	}
-	b.mu.Unlock() // release lock before shutdown
+	b.mu.Unlock()
+
+	b.webClient.mu.Lock() // webClient struct uses its own mutext
```

Comment:
We could potentially move `b.webClientListeners` into `b.webClient` as well and have that use `b.webClient.mu` as well...

Reply:
yeah, I kinda like that

Reply:
Oh actually, adding a new web client listener requires the netmap, which uses the b.mu lock anyways. I'll leave where it is for now, and we can move in a follow-up, although having to obtain both locks on creation much not actually be worth it.